File: android/lib/core/clipboard_manager.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';

import 'clipboard_adapter.dart';
import 'clipboard_model.dart';
import 'device_identity_manager.dart';
import 'lru_ring_cache.dart';
import 'protocol_frame.dart';
import 'session_manager.dart';

class ClipboardManager {
  final DeviceIdentity localIdentity;
  final ClipboardAdapter adapter;
  final LruRingCache cache;
  final List<SessionConnection> _activeConnections = [];

  final StreamController<ClipboardEventMessage> _clipboardController =
      StreamController<ClipboardEventMessage>.broadcast();
  final StreamController<String> _urlHandoffController =
      StreamController<String>.broadcast();

  StreamSubscription<String>? _adapterSubscription;
  String? _lastAppliedRemoteContent;
  bool autoSync = true;
  bool autoOpenUrls = false;

  Stream<ClipboardEventMessage> get clipboardStream =>
      _clipboardController.stream;
  Stream<String> get urlHandoffStream => _urlHandoffController.stream;

  ClipboardManager({
    required this.localIdentity,
    required this.adapter,
    LruRingCache? cache,
  }) : cache = cache ?? LruRingCache(100) {
    _adapterSubscription = adapter.textStream.listen(_onLocalClipboardChanged);
  }

  void registerConnection(SessionConnection connection) {
    if (!_activeConnections.contains(connection)) {
      _activeConnections.add(connection);
    }
  }

  void unregisterConnection(SessionConnection connection) {
    _activeConnections.remove(connection);
  }

  void _onLocalClipboardChanged(String text) {
    if (!autoSync || text.isEmpty) return;

    // Anti-echo: if the clipboard change matches what we just received from a peer, ignore it
    if (text == _lastAppliedRemoteContent) return;

    final eventId = generateUuid();
    cache.add(eventId);

    final msg = ClipboardEventMessage(
      eventId: eventId,
      originDeviceId: localIdentity.deviceId,
      timestamp: DateTime.now().millisecondsSinceEpoch,
      contentType: 'text/plain',
      content: text,
    );

    broadcastClipboardEvent(msg);
  }

  void broadcastClipboardEvent(ClipboardEventMessage msg) {
    final active = _activeConnections
        .where((c) => c.state == SessionState.connected)
        .toList();

    for (final conn in active) {
      conn.sendJson(MessageType.clipboardEvent, msg.toJson());
    }
  }

  Future<bool> handleIncomingFrame(
    SessionConnection? conn,
    ProtocolFrame frame,
  ) async {
    if (frame.type != MessageType.clipboardEvent) return false;

    ClipboardEventMessage msg;
    try {
      final json =
          jsonDecode(utf8.decode(frame.payload)) as Map<String, dynamic>;
      msg = ClipboardEventMessage.fromJson(json);
    } catch (_) {
      return false;
    }

    // 1. Anti-Loop Rule 1: Drop if originating from self
    if (msg.originDeviceId.toLowerCase() ==
        localIdentity.deviceId.toLowerCase()) {
      return false;
    }

    // 2. Anti-Loop Rule 2: Drop if already received (in LRU 100 cache)
    if (cache.contains(msg.eventId)) {
      return false;
    }

    // Add to ring cache
    cache.add(msg.eventId);

    // Apply to local clipboard
    _lastAppliedRemoteContent = msg.content;
    await adapter.setText(msg.content);

    _clipboardController.add(msg);

    // Check for URL handoff
    if (isUrl(msg.content)) {
      _urlHandoffController.add(msg.content.trim());
    }

    return true;
  }

  static bool isUrl(String? text) {
    if (text == null || text.trim().isEmpty) return false;
    final trimmed = text.trim();
    final uri = Uri.tryParse(trimmed);
    if (uri != null &&
        uri.hasScheme &&
        (uri.scheme == 'http' || uri.scheme == 'https')) {
      return uri.host.isNotEmpty;
    }
    return false;
  }

  static String generateUuid() {
    final random = Random.secure();
    final bytes = List<int>.generate(16, (_) => random.nextInt(256));
    bytes[6] = (bytes[6] & 0x0f) | 0x40; // version 4
    bytes[8] = (bytes[8] & 0x3f) | 0x80; // variant RFC4122
    final hex = bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
    return '${hex.substring(0, 8)}-${hex.substring(8, 12)}-${hex.substring(12, 16)}-${hex.substring(16, 20)}-${hex.substring(20, 32)}';
  }

  void dispose() {
    _adapterSubscription?.cancel();
    _clipboardController.close();
    _urlHandoffController.close();
    adapter.dispose();
  }
}

File: windows/Hinge.Core/ClipboardManager.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace Hinge.Core;

public class ClipboardManager : IDisposable
{
    private readonly DeviceIdentity _localIdentity;
    private readonly IClipboardAdapter _clipboardAdapter;
    private readonly LruRingCache _cache;
    private readonly List<SessionConnection> _activeConnections = new();
    private readonly object _connectionLock = new();

    private string? _lastAppliedRemoteContent;
    private bool _disposed;

    public bool AutoSync { get; set; } = true;
    public bool AutoOpenUrls { get; set; } = false;
    public LruRingCache Cache => _cache;
    public IClipboardAdapter Adapter => _clipboardAdapter;

    public event EventHandler<ClipboardEventMessage>? ClipboardReceived;
    public event EventHandler<string>? UrlHandoffReceived;

    public ClipboardManager(
        DeviceIdentity localIdentity,
        IClipboardAdapter clipboardAdapter,
        LruRingCache? cache = null)
    {
        _localIdentity = localIdentity ?? throw new ArgumentNullException(nameof(localIdentity));
        _clipboardAdapter = clipboardAdapter ?? throw new ArgumentNullException(nameof(clipboardAdapter));
        _cache = cache ?? new LruRingCache(100);

        _clipboardAdapter.TextChanged += OnLocalClipboardChanged;
    }

    public void RegisterConnection(SessionConnection connection)
    {
        lock (_connectionLock)
        {
            if (!_activeConnections.Contains(connection))
            {
                _activeConnections.Add(connection);
            }
        }
    }

    public void UnregisterConnection(SessionConnection connection)
    {
        lock (_connectionLock)
        {
            _activeConnections.Remove(connection);
        }
    }

    private void OnLocalClipboardChanged(object? sender, string text)
    {
        if (!AutoSync || string.IsNullOrEmpty(text)) return;

        // Anti-echo: if the clipboard change is the exact content we just applied from a remote peer, ignore it
        if (string.Equals(text, _lastAppliedRemoteContent, StringComparison.Ordinal))
        {
            return;
        }

        var eventId = Guid.NewGuid().ToString();
        _cache.Add(eventId);

        var msg = new ClipboardEventMessage
        {
            EventId = eventId,
            OriginDeviceId = _localIdentity.DeviceId,
            Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            ContentType = "text/plain",
            Content = text
        };

        BroadcastClipboardEvent(msg);
    }

    public void BroadcastClipboardEvent(ClipboardEventMessage msg)
    {
        List<SessionConnection> conns;
        lock (_connectionLock)
        {
            conns = _activeConnections.Where(c => c.State == SessionState.Connected).ToList();
        }

        foreach (var conn in conns)
        {
            _ = conn.SendJsonAsync(MessageType.ClipboardEvent, msg);
        }
    }

    public async Task<bool> HandleIncomingFrameAsync(SessionConnection conn, ProtocolFrame frame)
    {
        if (frame.Type != MessageType.ClipboardEvent) return false;

        ClipboardEventMessage? msg;
        try
        {
            msg = JsonSerializer.Deserialize<ClipboardEventMessage>(frame.Payload);
        }
        catch
        {
            return false;
        }

        if (msg == null) return false;

        // 1. Anti-Loop Rule 1: Drop if originating from self
        if (string.Equals(msg.OriginDeviceId, _localIdentity.DeviceId, StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        // 2. Anti-Loop Rule 2: Drop if already processed (exists in LRU 100 cache)
        if (_cache.Contains(msg.EventId))
        {
            return false;
        }

        // Add to ring cache
        _cache.Add(msg.EventId);

        // Apply to local clipboard
        _lastAppliedRemoteContent = msg.Content;
        await _clipboardAdapter.SetTextAsync(msg.Content);

        ClipboardReceived?.Invoke(this, msg);

        // Check for URL handoff
        if (IsUrl(msg.Content))
        {
            UrlHandoffReceived?.Invoke(this, msg.Content.Trim());
        }

        return true;
    }

    public static bool IsUrl(string? text)
    {
        if (string.IsNullOrWhiteSpace(text)) return false;
        text = text.Trim();
        if (Uri.TryCreate(text, UriKind.Absolute, out var uri))
        {
            return uri.Scheme == Uri.UriSchemeHttp || uri.Scheme == Uri.UriSchemeHttps;
        }
        return false;
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        _clipboardAdapter.TextChanged -= OnLocalClipboardChanged;
        _clipboardAdapter.Dispose();
    }
}
